src/pages/SmartSwatch.js

```diff
@@ -18,54 +20,114 @@ import {
   Text,
   Link,
   useColorMode,
+  useClipboard,
 } from '@chakra-ui/core';
 import { FaPalette, FaHeart } from "react-icons/fa";
 
 import { Swatch } from '../components/Swatch';
 import { Logo } from '../components/Logo';
 
+const PRESET_COLORS = [
+  '#C70833',
+  '#E53E3E',
+  '#F6AD55',
+  '#F6E05E',
+  '#48BB78',
+  '#4FD1C5',
+  '#4299E1',
+  '#0BC5EA',
+  '#805AD5',
+  '#D53F8C',
+  '#718096',
+];
+const DEFAULT_USER_COLOR_STRING = PRESET_COLORS[0];
+const UNKNOWN_USER_COLOR_STRING = '#000';
+const LIGHTNESS_MAP = [0.95, 0.85, 0.75, 0.65, 0.55, 0.45, 0.35, 0.25, 0.15, 0.05];
+const SATURATION_MAP = [0.32, 0.16, 0.08, 0.04, 0, 0, 0.04, 0.08, 0.16, 0.32];
+const HUE_MAP = [0, 4, 8, 12, 16, 20, 24, 28, 32, 36];
+
+const getColorStringFromHash = () => {
+  const sanitizedHash = window.location.hash.replace('#', '');
+  return chroma.valid(sanitizedHash) ? chroma(sanitizedHash).name() : DEFAULT_USER_COLOR_STRING;
+}
+
+const getUserColorChroma = (colorString, fallbackValue = UNKNOWN_USER_COLOR_STRING) => chroma.valid(colorString)
+  ? chroma(colorString)
+  : chroma(fallbackValue);
+
 export const SmartSwatch = () => {
-  const [userColorInput, setUserColorInput] = useState('#C70833');
+  const [userColorInput, setUserColorInput] = useState(getColorStringFromHash());
+  const [userColorChroma, setUserColorChroma] = useState(getUserColorChroma(userColorInput.trim()));
   const { colorMode, toggleColorMode } = useColorMode();
 
-  const lightnessMap = [0.95, 0.85, 0.75, 0.65, 0.55, 0.45, 0.35, 0.25, 0.15, 0.05];
-  const saturationMap = [0.32, 0.16, 0.08, 0.04, 0, 0, 0.04, 0.08, 0.16, 0.32];
-  const hueMap = [0, 4, 8, 12, 16, 20, 24, 28, 32, 36];
+  const { onCopy, hasCopied } = useClipboard(window.location);
+
+  React.useEffect(function updateColorChromaFromInputEffect() {
+    const trimmedColor = userColorInput.trim();
+    setUserColorChroma(getUserColorChroma(trimmedColor));
+
 
-  const userColor = chroma.valid(userColorInput.trim())
-    ? chroma(userColorInput.trim())
-    : chroma('#000');
+    // Push new hash only if:
+    //  - represents a valid chroma color
+    //  - it's not the same color already in the hash so we avoid duplicated history states
+    if (chroma.valid(trimmedColor) && userColorInput !== getColorStringFromHash()) {
+      const newHash = trimmedColor.startsWith('#') ? trimmedColor : `#${trimmedColor}`;
 
-  const lightnessGoal = userColor.get('hsl.l');
-  const closestLightness = lightnessMap.reduce((prev, curr) =>
+      // Using pushState rather than setting new hash directly through window.history.hash
+      // because pushState never causes a hashchange event to be fired.
+      window.history.pushState(null, null, newHash);
+    }
+  }, [userColorInput]);
+
+  React.useEffect(function setupHashChangeEventListenerEffect() {
+    const updateColorInput = () => {
+      setUserColorInput(getColorStringFromHash())
```

Comment:
I think you should debounce this because else it will update the hash at every keyboard input 😢 
(and you can use anonymous arrow function in useEffect 😉)

Example (not tested 😄)
```js
React.useEffect(() => {
    let timer;
    const updateColorInput = () => {
      setTimeout(() => { 
        timer = setUserColorInput(getColorStringFromHash())
      }, 500);
    };

    // Listening to hashchange event, so combined with pushState method we make
    // sure to only listen to hash changes because the user goes back/forward
    // in browser history, and we discard hash changes because the user picked
    // a new color.
    window.addEventListener('hashchange', updateColorInput, false);
    return () => {
      window.removeEventListener('hashchange', updateColorInput);
      clearTimeout(timer);
    };
  }, []);
```

Reply:
![May-25-2020 18-14-24](https://user-images.githubusercontent.com/9749061/82829488-aef75200-9eb3-11ea-8aff-94118f9833d4.gif)

Reply:
With debounce (500ms or maybe 1sec) it can solve this problem I think

Reply:
Sure, should be easy to add. But I'm not sure if this is what you mean. This piece of code is in charge of setting the input value from the hash, so it's only triggered when you go back/forward in the browser.

But you mentioned "it will update the hash at every keyboard input". That's not handled here, but in `updateColorChromaFromInputEffect` and it's not changing the the hash at every keyboard input but only when it's a valid color. Anyway, I see in your gif that it has weird behavior with browser history navigation so I can add debounce to both restore color from hash and update hash from input.

Reply:
The problem is with "in between" values like `#4299` or `#1345`

Reply:
Because of opacity :(

Reply:
#1234 will cast as #11223344 (where 44 is for alpha)

Reply:
(I don't think that the app should allow alpha values by the way, but never look how to fix that)

Reply:
"But I'm not sure if this is what you mean." oh wrong place in fact

Reply:
Yes the debounce should be in the hash update, my bad

Reply:
Updated, let me know what you think.